### PR TITLE
refact kubectl Factory make it interface

### PR DESCRIPTION
**Pull request by adohe-zz**
refactor kubectl to make Factory interface. @kubernetes/kubectl

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33180)

### CODE REVIEW

**Comment by deads2k** [comment]
@fabianofranz This ripples downstream.

**Comment by deads2k** [comment]
I'm in favor of the change, but wow that's big :)

**Comment by MHBauer** [comment]
related to #31702 ?

**Comment by deads2k** [comment]
> related to #31702 ?

Looks like it.

**Comment by smarterclayton** [comment]
Please split the `*cmdutil.Factory` change in each command into its own commit - that will make this much easier to review.

**Comment by AdoHe** [comment]
yes, sure, I will split this into separate commits.

**Comment by AdoHe** [comment]
@fabianofranz As we have discussed on email, we will propose a new SIG for kubectl, which gives us a place to discuss the CLI architecture, like this one. So here I want to make sure this is the correct way. I will get this merged asap. @pwittrock ptal.

File: pkg/kubectl/cmd/cmd_test.go [hunk 1]
```diff
@@ -209,7 +213,12 @@ type testFactory struct {
 	Err          error
 }
 
-func NewTestFactory() (*cmdutil.Factory, *testFactory, runtime.Codec, runtime.NegotiatedSerializer) {
+type fakeFactory struct {
+	tf    *testFactory
```
Comment (soltysh) [inline pkg/kubectl/cmd/cmd_test.go]: Would it make sense to have a general fakeFactory for tests, that will have all the methods implemented, with empty bodies, of course. Then in commands you'd be embedding it and just providing an implementation to whichever method you need. This should simplify testing, imho.

Comment (AdoHe) [inline pkg/kubectl/cmd/cmd_test.go]: @soltysh a general fakeFactory with all methods implemented with empty bodies? what if several commands want to use a same implementation of some methods?

Comment (soltysh) [inline pkg/kubectl/cmd/cmd_test.go]: That shouldn't be a problem for several commands re-using those.

File: pkg/kubectl/cmd/cmd_test.go [hunk 2]
```diff
@@ -219,167 +228,333 @@ func NewTestFactory() (*cmdutil.Factory, *testFactory, runtime.Codec, runtime.Ne
...
+}
+
+func (f *fakeFactory) BindExternalFlags(flags *pflag.FlagSet) {
+}
+
+func (f *fakeFactory) PrintObject(cmd *cobra.Command, mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error {
+	return nil
+}
+
+func (f *fakeFactory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMapping, withNamespace bool) (kubectl.ResourcePrinter, error) {
```
Comment (k82cn) [inline pkg/kubectl/cmd/cmd_test.go]: Should we return `t.tf.Printer` ?

File: pkg/kubectl/cmd/cmd_test.go [hunk 3]
```diff
@@ -219,167 +228,333 @@ func NewTestFactory() (*cmdutil.Factory, *testFactory, runtime.Codec, runtime.Ne
...
+	return nil, nil
+}
+
+func (f *fakeFactory) NewBuilder() *resource.Builder {
+	return nil
```
Comment (k82cn) [inline pkg/kubectl/cmd/cmd_test.go]: The `DrainOptions.SetupDrain` in `cmd/drain.go` will use `NewBuidler` of factory, and the default implementation of `NewFactory` will take `Factory::Object`'s return value as parameters. That's what I mean in slack.

But we're luck, it seems `DrainOptions.SetupDrain` is not used by anyone :).

**Review by k82cn — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/cmd_test.go (hunk 3 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/cmd_test.go]: we still need to be careful about this.

File: pkg/kubectl/cmd/cmd_test.go (hunk 2 above)
Comment (deads2k) [inline pkg/kubectl/cmd/cmd_test.go]: > Should we return t.tf.Printer

I think we should keep this pull targeted.  Switch them all to methods on the `Factory` interface and then change signatures later.  This pull is already massive.

File: pkg/kubectl/cmd/util/factory.go [hunk 4]
```diff
@@ -74,90 +74,108 @@ const (
...
 // commands are decoupled from the factory).
-type Factory struct {
-	clients *ClientCache
-	flags   *pflag.FlagSet
-
-	// Returns interfaces for dealing with arbitrary runtime.Objects.
-	Object func() (meta.RESTMapper, runtime.ObjectTyper)
+type Factory interface {
+	// Returns the inner flag set
+	Flags() *pflag.FlagSet
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: This used to be private, right?  Can we keep it private?

File: pkg/kubectl/cmd/util/factory.go [hunk 5]
```diff
@@ -74,90 +74,108 @@ const (
...
 	// of their choice.
-	EditorEnvs func() []string
+	EditorEnvs() []string
 	// PrintObjectSpecificMessage prints object-specific messages on the provided writer
-	PrintObjectSpecificMessage func(obj runtime.Object, out io.Writer)
+	PrintObjectSpecificMessage(obj runtime.Object, out io.Writer)
+	// Command will stringify and return all environment arguments ie. a command run by a client
+	// using the factory.
+	// TODO: We need to filter out stuff like secrets.
+	Command() string
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: Why have these been added?

File: pkg/kubectl/cmd/util/factory.go [hunk 6]
```diff
@@ -74,90 +74,108 @@ const (
...
+	// BindFlags adds any flags that are common to all kubectl sub commands.
+	BindFlags(flags *pflag.FlagSet)
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: If these weren't exposed before, lets keep them hidden.  I don't think we want to encourage access patterns like this.

File: pkg/kubectl/cmd/util/factory.go [hunk 7]
```diff
@@ -74,90 +74,108 @@ const (
...
+	// BindExternalFlags adds any flags defined by external projects (not part of pflags)
+	BindExternalFlags(flags *pflag.FlagSet)
+	// PrintObject prints an api object given command line flags to modify the output format
+	PrintObject(cmd *cobra.Command, mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error
```
Comment (deads2k) [inline pkg/kubectl/cmd/util/factory.go]: please remove/hide.  We don't want to the factory accepting the command.  We want specific structs passing the information we need and all of these look be net-new

**Comment by deads2k** [comment]
Mind squashing into 
 1. update factory.
 1. create test factory
 1. mechanical command updates? - looks like these were mostly just removing the `*`.

I'd like to see the interface slimmed down a little bit.  If that turns out to require rippling changes, just make follow-up issues.  This is a nice improvement over where we are now.

**Comment by deads2k** [comment]
After squash and green, I'd merge this as-is and we can iterate.

I bumped the priority, once its green, we want this to beat any other `kubectl` changes.

**Comment by AdoHe** [comment]
```
3. mechanical command updates? - looks like these were mostly just removing the *.
```
yes, so I would like to merge these into one commit, and one commit one command would leave so many commits tests failed.

**Comment by AdoHe** [comment]
I just go back to this issue, will fix it tomorrow.

**Comment by deads2k** [comment]
Build issues.  If you think you've got a shot at finishing it up today, ask for a hold on https://github.com/kubernetes/kubernetes/pull/34199 and untag.

**Comment by AdoHe** [comment]
the build failed because I just update the `Factory` struct in this commit, will send other commits to update various commands.

File: pkg/kubectl/cmd/util/factory.go (hunk 5 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/util/factory.go]: @deads2k originally `Factory` struct has these methods, and they are exported, so other packages can use. If we make `Factory` interface, and not add these methods to the interface definition, other packages which use this interface can not use theses methods anymore.

File: pkg/kubectl/cmd/util/factory.go (hunk 6 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/util/factory.go]: @deads2k actually these are methods of `Factory` struct before, and they are exported. The problem is why we add these methods as struct methods, and others as struct fields.

**Comment by AdoHe** [comment]
@deads2k @fabianofranz @pwittrock I just rebased and fixed the failed tests, please take a look, and give your feedback, I would hope we can get this merged asap. This is just a beginning of `Factory` refactor, eventually we will make `Factory` a collection of interfaces, and finally we will implement something we discussed in #31173 #7311

File: pkg/kubectl/cmd/annotate.go [hunk 8]
```diff
@@ -54,7 +54,7 @@ type AnnotateOptions struct {
 	newAnnotations    map[string]string
 	removeAnnotations []string
 
-	// Common shared fields
```
Comment (kargakis) [inline pkg/kubectl/cmd/annotate.go]: shared is correct here

File: pkg/kubectl/cmd/drain.go [hunk 9]
```diff
@@ -419,6 +419,7 @@ func (o *DrainOptions) RunCordonOrUncordon(desired bool) error {
 			unsched.SetBool(desired)
 			_, err := helper.Replace(cmdNamespace, o.nodeInfo.Name, true, o.nodeInfo.Object)
 			if err != nil {
+				fmt.Println("get error here")
```
Comment (kargakis) [inline pkg/kubectl/cmd/drain.go]: remove this?

File: pkg/kubectl/cmd/util/factory.go [hunk 10]
```diff
@@ -77,87 +77,107 @@ const (
...
+	// DefaultResourceFilterFunc returns a collection of FilterFuncs suitable for filtering specific resource types.
+	DefaultResourceFilterFunc() kubectl.Filters
+
+	// PrintObject prints an api object given command line flags to modify the output format
+	PrintObject(cmd *cobra.Command, mapper meta.RESTMapper, obj runtime.Object, out io.Writer) error
+	// PrinterForMapping returns a printer suitable for displaying the provided resource type.
+	// Requires that printer flags have been added to cmd (see AddPrinterFlags).
+	PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMapping, withNamespace bool) (kubectl.ResourcePrinter, error)
+	// One stop shopping for a Builder
+	NewBuilder() *resource.Builder
```
Comment (kargakis) [inline pkg/kubectl/cmd/util/factory.go]: Nice

File: pkg/kubectl/cmd/drain.go (hunk 9 above)
Comment (AdoHe) [inline pkg/kubectl/cmd/drain.go]: some stupid error, :(

**Comment by AdoHe** [comment]
@deads2k @fabianofranz @pwittrock @kargakis rebased and update, ptal.

**Comment by deads2k** [comment]
> @deads2k @fabianofranz @pwittrock @kargakis rebased and update, ptal.

I don't see an obvious, "this doesn't work", so I'm tagging this as-is.  This breaks the back of the problem and I'd like to merge it and take everything else as a follow-up unless someone finds a "this breaks `kubectl`" kind of bug.

**Comment by kargakis** [comment]
Agreed.

**Comment by AdoHe** [comment]
wow~ so glad to see this thanks @deads2k @kargakis @fabianofranz @pwittrock

**Pull request merged**